graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/migration/CassandraMigrator.java

```diff
@@ -0,0 +1,174 @@
+              actualType,
+              CassandraSchemaHelper::compare,
+              CassandraSchemaHelper::buildCreateQuery,
+              CassandraSchemaHelper::buildDropQuery,
+              CassandraSchemaHelper::buildAddColumnQuery,
+              "UDT",
+              queries,
+              errors);
+          break;
+        default:
+          throw new AssertionError("Unexpected target " + entity.getTarget());
+      }
+    }
+    if (!errors.isEmpty()) {
+      throw GraphqlErrorException.newErrorException()
+          .message(
+              String.format(
+                  "The schema you provided can't be mapped to the current CQL data model."
+                      + "Consider using a different migration strategy (current: %s). "
+                      + "See details in `extensions.migrationErrors` below.",
+                  strategy))
+          .extensions(ImmutableMap.of("migrationErrors", errors))
+          .build();
+    }
+    return ImmutableList.copyOf(queries);
+  }
+
+  private <T extends SchemaEntity> void compute(
+      T expected,
+      T actual,
+      BiFunction<T, T, List<Difference>> comparator,
+      BiFunction<T, DataStore, MigrationQuery> createBuilder,
+      BiFunction<T, DataStore, MigrationQuery> dropBuilder,
+      CassandraSchemaHelper.AddColumnBuilder<T> addColumnBuilder,
+      String entityType,
+      Set<MigrationQuery> queries,
+      List<String> errors) {
+    if (strategy == MigrationStrategy.DROP_AND_RECREATE_ALL) {
+      // We don't need to compare the schema, just override everything.
+      queries.add(dropBuilder.apply(expected, dataStore));
+      queries.add(createBuilder.apply(expected, dataStore));
+    } else if (actual == null) {
+      // The table/UDT does not exist
+      switch (strategy) {
+        case USE_EXISTING:
+          errors.add(String.format("Missing %s %s", entityType, expected.name()));
+          break;
+        case ADD_MISSING_TABLES:
+        case ADD_MISSING_TABLES_AND_COLUMNS:
+        case DROP_AND_RECREATE_IF_MISMATCH:
+          queries.add(createBuilder.apply(expected, dataStore));
+          break;
+        default:
+          throw new AssertionError("Unexpected strategy " + strategy);
+      }
+    } else {
+      // The table/UDT exists, compare with our model
+      List<Difference> differences = comparator.apply(expected, actual);
+      if (!differences.isEmpty()) {
+        switch (strategy) {
+          case USE_EXISTING:
+          case ADD_MISSING_TABLES:
+            errors.addAll(
+                differences.stream().map(Difference::toString).collect(Collectors.toList()));
+            break;
+          case ADD_MISSING_TABLES_AND_COLUMNS:
+            List<Difference> blockers =
+                differences.stream().filter(d -> !isAddableColumn(d)).collect(Collectors.toList());
+            if (blockers.isEmpty()) {
+              for (Difference difference : differences) {
+                assert isAddableColumn(difference);
```

Comment:
why assert? Should we check and throw an exception if it does not match?

Reply:
Because I know at this point that this is true, otherwise `blockers` would not be empty.

The only way this could fail is a programming error. This is more about documenting it, and protecting against future refactorings (e.g. if we change the filter condition in `blockers` but forget to revisit this). I feel that's a good fit for assertions.